### ISSUE TITLE
[FG:InPlacePodVerticalScaling] Add support in containerd for returning Resources in ContainerStatus CRI API

### ISSUE THREAD

**Issue by vinaykul** [kubernetes/kubernetes#107911]
### What would you like to be added?

[In-place Pod Vertical Scaling feature](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources) allows for resizing of CPU and memory resources of running pods without having to restart the pod and its containers.

This is achieved by making the Container.Resources field mutable and allowing the API user to use it to express **desired resources** for the pod's containers. The actual resources actuated on the pod's containers are shown in a newly added ContainerStatus.Resources field in PodStatus. The actual resources that has been actuated on the container needs to come from the runtime in order for kubelet to close the loop on achieving "actual resources == desired resources"

This requires support from container runtimes such as containerd. The runtime needs to handle an extension to the [ContainerStatus CRI API](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2273-kubelet-container-resources-cri-api-changes#api-changes) and return actual resources in effect for the container in the Resources field as part of the response message to ContainerStatus API.

A reference example implementation for now deprecated dockershim can be found in [this PR](https://github.com/kubernetes/kubernetes/pull/92127/files#diff-bd4bca8e04fa909b829b17bb54dc42d900dbff59e662870b649651f0a8d165a3R439)

### Why is this needed?

Without this, we cannot know what actual resources have been actuated on a pod's containers, and hence cannot close the loop on achieving "actual state == desired state" for pod's resources.

This is a beta blocker for In-place Pod Vertical Scaling [feature](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources)

**Comment by vinaykul** [comment]
/assign @ruiwen-zhao 
/assign @mikebrow 
Ruiwen has offered to help with this. Thanks a lot Ruiwen!

**Comment by adisky** [comment]
I can also help with this!!

**Comment by ruiwen-zhao** [comment]
Looking at the containerd code, I think we just need to add to containerd's [container_status.go](https://github.com/containerd/containerd/blob/f0a32c66dad1e9de716c9960af806105d691cd78/pkg/cri/server/container_status.go) something like this:

```
runtimeSpec, err := container.Container.Spec(ctx)
...
LogPath:     meta.LogPath,
Resources:   &runtime.ContainerResources{
  Linux: &runtime.LinuxContainerResources{
    CpuPeriod:          runtimeSpec.Linux.Resources.CPU.Period,
    CpuQuota:           runtimeSpec.Linux.Resources.CPU.Quota,
    CpuShares:          runtimeSpec.Linux.Resources.CPU.Shares,
    MemoryLimitInBytes: runtimeSpec.Linux.Resources.Memory.Limit,
    OomScoreAdj:        runtimeSpec.Process.OOMScoreAdj,
    CpusetCpus:         runtimeSpec.Linux.Resources.CPU.Cpus,
    CpusetMems:         runtimeSpec.Linux.Resources.CPU.Mems,
  },
},
``` 

This work is currently blocked by: a) cri-api updated with `Resources` field, which is part of https://github.com/kubernetes/kubernetes/pull/102884, and b) containerd updating the cri-api. 

Lemme know if I have misunderstood anything. Thanks!

**Comment by vinaykul** [comment]
@ruiwen-zhao I did a quick scan. It looks like the right place to add support. Please verify container.Container.Spec() returns the resource values that are in-effect on the container (actual cgroup values).

You can try it out using this PR if there's a way to pick up runtime/v1alpha2 locally. But yes you will need to wait for it to merge and CRI update to be picked up by containerd before any commits to containerd can go in. @mikebrow can guide you better about the processes.

Our E2E test ( test/e2e/node/pod_resize.go ) in this PR can help you verify when 'featureGatePostAlpha' is true. This test currently verifies cgroup values of containerd are as expected but skips verifying Status.Resources tally up (until beta)

I run it with local cluster using containerd as follows:
```bash
# CONTAINER_RUNTIME=remote CONTAINER_RUNTIME_ENDPOINT="unix:///run/containerd/containerd.sock" sudo -E ./hack/local-up-cluster.sh 
...
```
```bash
# kubetest --test --test_args="--ginkgo.focus=InPlacePodVerticalScaling" --provider=local
```

**Comment by ruiwen-zhao** [comment]
> Please verify container.Container.Spec() returns the resource values that are in-effect on the container (actual cgroup values).

According to [runtime-spec](https://github.com/opencontainers/runtime-spec/blob/main/specs-go/config.go#L163), the `Resources` contain cgroup information for handling resource constraints for the container. So I think they are the values we are looking to get here.

**Comment by ruiwen-zhao** [comment]
Created a draft PR on containerd https://github.com/containerd/containerd/pull/6517. Unit test now passes but I will probably need the CRI change to be merged for e2e test.

**Comment by ruiwen-zhao** [comment]
Containerd PR is now merged: https://github.com/containerd/containerd/pull/6517

Next step is to update the containerd dependency on k/k.

**Comment by vinaykul** [comment]
> Containerd PR is now merged: [containerd/containerd#6517](https://github.com/containerd/containerd/pull/6517)
> 
> Next step is to update the containerd dependency on k/k.

@ruiwen-zhao Thank you very much for driving this!

**Comment by ruiwen-zhao** [comment]
The containerd support https://github.com/containerd/containerd/pull/6517 is done and k/k now has e2e tests using the containerd support. Closing.

/close
